### PR TITLE
Port dev(ambilight) branch to v17

### PR DESCRIPTION
**Pull request by Moyster**
- Used @koying dev branch to get rid of "media info from log" and the
framerate hack
- useLegacyApi + live TV support (@fritsch)
see : https://github.com/koying/script.kodi.hue.ambilight/commits/develop

This works for me, should this be branch "v17" ?

### CODE REVIEW

**Comment by fritsch** [comment]
Wasn't it possible to pick the original patches from their remote branches to keep history?

**Comment by fritsch** [comment]
And now pick the stuff and afterwards do a rebase and force push - then you are set.

**Comment by Moyster** [comment]
I'm where I started fiddling around to make it work on v17, importmediainfofromlog is giving errors so I just removed it.  
Adding the useLegacyapi was pretty simple but didn't apply correctly on " def run():" (the while part)
Those 3 commits helped me sort the While loop :   
- create player before loop : https://github.com/koying/script.kodi.hue.ambilight/commit/2e23a8fad5adc96b3597f60920c174cd959841f6  
-  refactor waitForAbort : https://github.com/koying/script.kodi.hue.ambilight/commit/93a6809ec4547c1fa255a1520873a11f183847be
- refactor ambient / remove framerate hack : https://github.com/koying/script.kodi.hue.ambilight/commit/a6db47b70c0f304b6f1a0192e9da1ce2d28378b4

But they don't merge correctly and the while loop isn't functional (lights just dim on play)
After fixing the while loop, every commits from "https://github.com/koying/script.kodi.hue.ambilight/commit/d909c1b954c0ec0f128fbbbe835d5fafa92b3783" til the last one should merge correctly tho.

Should I just apply the minimum to make it work on v17 and then merge the superflous commits ?

**Comment by fritsch** [comment]
You need to ask @mpolednik how he wants it. As he is the upstream no merge commits are allowed, rebase ontop of this dev branch, so that he can easily pick your patches.

**Comment by fritsch** [comment]
git remote add mpolednik https://github.com/mpolednik/script.kodi.hue.ambilight.git
Backup what you have: git branch mybackupofdevelop
Now rebase on his develop branch:
git rebase -i mpolednik/develop

Now drop all the commits that are obsolete (remove the complete line). When done:

git push -f origin develop

Something went nuts?

git rebase --abort
git reset --hard mybackupofdevelop

and retry :-)

**Comment by mpolednik** [comment]
Please make sure it does not break v16, maybe with the upstream transition we could branch out v16 compatible version and keep develop (-> master) for v17. Not sure how people will like that, gonna take some time with that.

**Comment by fritsch** [comment]
https://github.com/mpolednik/script.kodi.hue.ambilight/pull/8/commits/73d2964d6c65f31c23a25b4f070538d70ecb3932 <- does not break compatibility
https://github.com/mpolednik/script.kodi.hue.ambilight/pull/8/commits/0f15bd96f4758805af9cc1e121c9f2a4ee8fcea8 <- does as this setting is named differently in v16, can be workarounded easily

**Comment by Moyster** [comment]
Is there a global variable to check kodi version just so it takes an if/else to make 0f15bd9 v16 & 17 compatible ? Or just don't commit 0f15bd9 to master ? :P

**Comment by fritsch** [comment]
If there was would I have implemented it as is? :-)

**Comment by Moyster** [comment]
Well, you said "can be workarounded easily", it might have been you being currently busy :-)

**Comment by Moyster** [comment]
Is this acceptable ?

**Comment by mpolednik** [comment]
It should be transparent to the user. Can you execute the JSONRPC command, check response for failure and act based on that?

**Comment by Moyster** [comment]
Currently, only your "develop" branch works (for v16 only),  
master is v17 only because of https://github.com/mpolednik/script.kodi.hue.ambilight/commit/0f15bd96f4758805af9cc1e121c9f2a4ee8fcea8 but far behind dev branch

except for that commit, my PR works on both versions, so I'm all for merging it to master & develop ( minus 0f15bd9 ) because it gives us a similar base to work on, independently from what Kodi version it should work on.

I guess it's up to you :-)
I'd like to add an option to disable ambilight when using live tv and @Anthrax-Leprosy-Pi seemed to be willing to contribute as well, it'd be great to "sort" the branches and have somewhere to send further PRs

**Comment by fritsch** [comment]
I cloned your dev branch, fixed the runtime error and succesfully running it on v17 now. If you adapt the json change to useLegacyApi then it will be fine.

**Comment by mpolednik** [comment]
Let me do some branching magic.

Let's work like this:
master -> old (let's possibly make it v16 only / v17 only but not actively work on it)
develop -> v16 "stable"
develop -> master with this PR in (optionally make it backwards compatible as a longer term goal)

**Comment by fritsch** [comment]
@mpolednik if the useLegacy change is adapted for the json field, this is not needed. Btw. v17 is the future. v16 is already dead from a dev point of view, there won't be any point releases anymore.

**Comment by mpolednik** [comment]
@fritsch I want to (selfishly! :) ) keep current develop as reference v16 branch and old code "just to be sure". This branching will allow us to make sure that v17 gets all the changes it needs in without having to fight backwards compatibility.

**Comment by fritsch** [comment]
Okay as you wish. A v17 branch will do it, obviously :-) and most stuff can just be cherry-picked. I wanted to avoid you further work and having to maintain two branches.

**Comment by mpolednik** [comment]
I suppose, wanted to make sure that nothing gets lost. Now we can freely work on master as the proper upstream branch and backport candy to v16.

**Comment by Moyster** [comment]
Yay, thanks guys :-)

**Comment by Anthrax-Leprosy-Pi** [comment]
Without any prior knowledge about the Kodi API I got some thoughts on
dealing with the breaking changes in the Kodi API...

Can't we just switch the usage of the legacy API based on the current Kodi
version? Therefore only master and develop branches would be needed and
compatibility is handled by the addon.

Just my two cents.

I'm going to play with it come weekend...

Cheers
Anthrax

Moyster <notifications@github.com> schrieb am Do., 16. Juni 2016, 17:21:

**Comment by Anthrax-Leprosy-Pi** [comment]
Quick Question:

Are you guys using any specific Python IDE or just hacking away in Notepad++?
I used Visual Studios with extensions and it screwed with the code formatting which resulted in A LOT OF CHANGES while committing...

;-)

**Comment by Moyster** [comment]
>> Can't we just switch the usage of the legacy API based on the current Kodi version?  
"if the useLegacy change is adapted for the json field, this is not needed." I wasn't around but I'm still wondering who's the laziest of us 4 not to do it :P

Since you said "notepad++", I assume you're on windows, you can look at Atom or keep going with notepad++, just make sure it uses spaces instead of tabs, furthermore @mpolednik said he wanted the code to stay "pep8 compliant" ish, at least from a far, see : https://www.python.org/dev/peps/pep-0008/

**Comment by AnthraxLeprosyPi** [comment]
Good evening everybody,

Wow, "learning" github & python has been a rollercoaster for me so far (spend the day installing sublime, atom, linter, pep8, flake8, autopep8, etc.) and hours later I arrived at PyCharm which gets the job done for me... 

Short version: not much progess code wise ;-)
 
So heres just a quick feedback:
I just tested the latest master branch and all my issues i previously had with the original published version of this addon are GONE! Lights turn on from off state automatically, response time is way, way better than before...

Short version: YOU GUYS ROCK!

bye
Anthrax
